src/software/simulated_tests/simulated_test_fixture.h

```diff
@@ -2,40 +2,165 @@
+     *
+     * @pre The robot IDs must not be duplicated and must not match the ID
+     * of any robot already on the specified team.
+     *
+     * @throws runtime_error if any of the given robot ids are duplicated, or a
+     * robot already exists on the specified team with one of the new IDs
+     *
+     * @param robots the robots to add
+     */
+    void addFriendlyRobots(const std::vector<RobotStateWithId>& robots);
+    void addEnemyRobots(const std::vector<RobotStateWithId>& robots);
+
+    /**
+     * Sets the goalie for the specified team. If this function is not called,
+     * the goalie will be set to the default ID of the DynamicParameters
+     *
+     * @param goalie_id The ID of the robot to be goalie
+     */
+    void setFriendlyGoalie(RobotId goalie_id);
+    void setEnemyGoalie(RobotId goalie_id);
+
+    /**
+     * Sets the play to run in the simulated test
+     *
+     * @param play_name The name of the play
+     */
+    void setPlay(const std::string& play_name);
+
+    /**
+     * Returns the field in the simulated test
+     *
+     * @return the field in the simulated test
+     */
+    Field field() const;
+
+   private:
+    /**
+     * A helper function that updates SensorFusion with the latest data from the Simulator
+     */
+    void updateSensorFusion();
+
+    /**
+     * Runs the given function validators and returns whether or not the
+     * FunctionValidators have completed (Note completed does not necessarily
+     * mean passed).
+     *
+     * @param function_validators The FunctionValidators to check
+     * @param continuous_function_validators The ContinuousFunctionValidators to check
+     *
+     * @return true if there is at least one FunctionValidator and all FunctionValidators
+     * have completed, and false otherwise
+     */
+    static bool validate(
+        std::vector<FunctionValidator>& function_validators,
+        std::vector<ContinuousFunctionValidator>& continuous_function_validators);
+
+    /**
+     * Puts the current thread to sleep until the difference between the current wall time
+     * and the wall_start_time is >= the difference between current_time and zero. This
+     * "synchronizes" wall time with simulation time by slowing down execution if needed.
+     *
+     * @param wall_start_time The time at which the simulated test started, in wall time
+     * @param current_time The current time in the simulated test
+     */
+    static void sleep(const std::chrono::steady_clock::time_point& wall_start_time,
+                      const Timestamp& current_time);
+
+    // The simulator needs to be a pointer so that we can destroy and re-create
+    // the object in the SetUp function. Because the simulator has no
+    // copy assignment operator, we have to make it a dynamically-allocated
+    // object so we can assign new instances to this variable
+    std::unique_ptr<Simulator> simulator;
+    SensorFusion sensor_fusion;
+    AI ai;
```

Comment:
missing comments for member variables. we should at least indicate that this is the AI and sensor fusion instance that we're using in our simulation.

Reply:
done